src/Rules/Regexp/RegularExpressionQuotingRule.php

```diff
@@ -0,0 +1,217 @@
+	{
+		return FuncCall::class;
+	}
+
+	public function processNode(Node $node, Scope $scope): array
+	{
+		if (!$node->name instanceof Node\Name) {
+			return [];
+		}
+
+		if (!$this->reflectionProvider->hasFunction($node->name, $scope)) {
+			return [];
+		}
+
+		$functionReflection = $this->reflectionProvider->getFunction($node->name, $scope);
+		if (
+			!in_array($functionReflection->getName(), [
+				'preg_match',
+				'preg_match_all',
+				'preg_filter',
+				'preg_grep',
+				'preg_replace',
+				'preg_replace_callback',
+				'preg_split',
+			], true)
+		) {
+			return [];
+		}
+
+		$normalizedArgs = $this->getNormalizedArgs($node, $scope, $functionReflection);
+		if ($normalizedArgs === null) {
+			return [];
+		}
+		if (!isset($normalizedArgs[0])) {
+			return [];
+		}
+		if (!$normalizedArgs[0]->value instanceof Concat) {
+			return [];
+		}
+
+		$patternDelimiter = $this->getDelimiterFromConcat($normalizedArgs[0]->value, $scope);
+		return $this->validateQuoteDelimiters($normalizedArgs[0]->value, $scope, $patternDelimiter);
+	}
+
+	/**
+	 * @return list<IdentifierRuleError>
+	 */
+	private function validateQuoteDelimiters(Concat $concat, Scope $scope, ?string $patternDelimiter): array
+	{
+		if ($patternDelimiter === null) {
+			return [];
+		}
+
+		$errors = [];
+		if (
+			$concat->left instanceof FuncCall
+			&& $concat->left->name instanceof Name
+			&& $concat->left->name->toLowerString() === 'preg_quote'
+		) {
+			$pregError = $this->validatePregQuote($concat->left, $scope, $patternDelimiter);
+			if ($pregError !== null) {
+				$errors[] = $pregError;
+			}
+		} elseif ($concat->left instanceof Concat) {
+			$errors = array_merge($errors, $this->validateQuoteDelimiters($concat->left, $scope, $patternDelimiter));
+		}
+
+		if (
+			$concat->right instanceof FuncCall
+			&& $concat->right->name instanceof Name
+			&& $concat->right->name->toLowerString() === 'preg_quote'
+		) {
+			$pregError = $this->validatePregQuote($concat->right, $scope, $patternDelimiter);
+			if ($pregError !== null) {
+				$errors[] = $pregError;
+			}
+		} elseif ($concat->right instanceof Concat) {
+			$errors = array_merge($errors, $this->validateQuoteDelimiters($concat->right, $scope, $patternDelimiter));
+		}
+
+		return $errors;
+	}
+
+	private function validatePregQuote(FuncCall $pregQuote, Scope $scope, string $patternDelimiter): ?IdentifierRuleError
+	{
+		if (!$pregQuote->name instanceof Node\Name) {
+			return null;
+		}
+
+		if (!$this->reflectionProvider->hasFunction($pregQuote->name, $scope)) {
+			return null;
+		}
+		$functionReflection = $this->reflectionProvider->getFunction($pregQuote->name, $scope);
+
+		$normalizedArgs = $this->getNormalizedArgs($pregQuote, $scope, $functionReflection);
+		if ($normalizedArgs === null) {
+			return null;
+		}
+
+		if (!isset($normalizedArgs[1])) {
+			return RuleErrorBuilder::message(sprintf('Call to preg_quote() is missing delimiter %s to be effective.', $patternDelimiter))
```

Comment:
This doesn't sound really true in context of `check for delimiters which get properly escaped by default`.

If the delimiter of the regex is one of those characters in the list, then a single-arg call to `preg_quote` is OK?

Reply:
this code was not reached for the case mentioned.

I refactored the code to make it more reasonable.